// reshuffle-slack-connector/dist/SlackEvents.js
"use strict";
exports.__esModule = true;
exports.SlackEvents = void 0;
var SlackEvents;
(function (SlackEvents) {
    SlackEvents["APP_REQUESTED"] = "app_requested";
    SlackEvents["APP_HOME_OPENED"] = "app_home_opened";
    SlackEvents["APP_MENTION"] = "app_mention";
    SlackEvents["APP_UNINSTALLED"] = "app_uninstalled";
    SlackEvents["CHANNEL_ARCHIVE"] = "channel_archive";
    SlackEvents["CHANNEL_CREATED"] = "channel_created";
    SlackEvents["CHANNEL_DELETED"] = "channel_deleted";
    SlackEvents["CHANNEL_HISTORY_CHANGED"] = "channel_history_changed";
    SlackEvents["CHANNEL_LEFT"] = "channel_left";
    SlackEvents["CHANNEL_RENAME"] = "channel_rename";
    SlackEvents["CHANNEL_SHARED"] = "channel_shared";
    SlackEvents["CHANNEL_UNARCHIVE"] = "channel_unarchive";
    SlackEvents["CHANNEL_UNSHARED"] = "channel_unshared";
    SlackEvents["DND_UPDATED"] = "dnd_updated";
    SlackEvents["DND_UPDATED_USER"] = "dnd_updated_user";
    SlackEvents["EMAIL_DOMAIN_CHANGED"] = "email_domain_changed";
    SlackEvents["EMOJI_CHANGED"] = "emoji_changed";
    SlackEvents["FILE_CHANGE"] = "file_change";
    SlackEvents["FILE_COMMENT_DELETED"] = "file_comment_deleted";
    SlackEvents["FILE_CREATED"] = "file_created";
    SlackEvents["FILE_DELETED"] = "file_deleted";
    SlackEvents["FILE_PUBLIC"] = "file_public";
    SlackEvents["FILE_SHARED"] = "file_shared";
    SlackEvents["FILE_UNSHARED"] = "file_unshared";
    SlackEvents["GRID_MIGRATION_FINISHED"] = "grid_migration_finished";
    SlackEvents["GRID_MIGRATION_STARTED"] = "grid_migration_started";
    SlackEvents["GROUP_ARCHIVE"] = "group_archive";
    SlackEvents["GROUP_CLOSE"] = "group_close";
    SlackEvents["GROUP_DELETED"] = "group_deleted";
    SlackEvents["GROUP_HISTORY_CHANGED"] = "group_history_changed";
    SlackEvents["GROUP_LEFT"] = "group_left";
    SlackEvents["GROUP_OPEN"] = "group_open";
    SlackEvents["GROUP_RENAME"] = "group_rename";
    SlackEvents["GROUP_UNARCHIVE"] = "group_unarchive";
    SlackEvents["IM_CLOSE"] = "im_close";
    SlackEvents["IM_CREATED"] = "im_created";
    SlackEvents["IM_HISTORY_CHANGED"] = "im_history_changed";
    SlackEvents["IM_OPEN"] = "im_open";
    SlackEvents["INVITE_REQUESTED"] = "invite_requested";
    SlackEvents["LINK_SHARED"] = "link_shared";
    SlackEvents["MEMBER_JOINED_CHANNEL"] = "member_joined_channel";
    SlackEvents["MEMBER_LEFT_CHANNEL"] = "member_left_channel";
    SlackEvents["MESSAGE"] = "message";
    SlackEvents["BOT_MESSAGE"] = "bot_message";
    SlackEvents["EKM_ACCESS_DENIED"] = "ekm_access_denied";
    SlackEvents["ME_MESSAGE"] = "me_message";
    SlackEvents["MESSAGE_CHANGED"] = "message_changed";
    SlackEvents["MESSAGE_DELETED"] = "message_deleted";
    SlackEvents["MESSAGE_REPLIED"] = "message_replied";
    SlackEvents["THREAD_BROADCAST"] = "thread_broadcast";
    SlackEvents["PIN_ADDED"] = "pin_added";
    SlackEvents["PIN_REMOVED"] = "pin_removed";
    SlackEvents["REACTION_ADDED"] = "reaction_added";
    SlackEvents["REACTION_REMOVED"] = "reaction_removed";
    SlackEvents["STAR_ADDED"] = "star_added";
    SlackEvents["STAR_REMOVED"] = "star_removed";
    SlackEvents["SUBTEAM_CREATED"] = "subteam_created";
    SlackEvents["SUBTEAM_MEMBERS_CHANGED"] = "subteam_members_changed";
    SlackEvents["SUBTEAM_SELF_ADDED"] = "subteam_self_added";
    SlackEvents["SUBTEAM_SELF_REMOVED"] = "subteam_self_removed";
    SlackEvents["SUBTEAM_UPDATED"] = "subteam_updated";
    SlackEvents["TEAM_DOMAIN_CHANGED"] = "team_domain_changed";
    SlackEvents["TEAM_JOIN"] = "team_join";
    SlackEvents["TEAM_RENAME"] = "team_rename";
    SlackEvents["TOKENS_REVOKED"] = "tokens_revoked";
    SlackEvents["USER_CHANGE"] = "user_change";
})(SlackEvents = exports.SlackEvents || (exports.SlackEvents = {}));
